src/overlays/actors/ovl_Obj_Warpstone/z_obj_warpstone.c

```diff
@@ -22,39 +30,156 @@ const ActorInit Obj_Warpstone_InitVars = {
-extern InitChainEntry D_80B9324C[];
-#endif
-
-#pragma GLOBAL_ASM("./asm/non_matchings/overlays/ovl_Obj_Warpstone_0x80B92B10/func_80B92B10.asm")
-
-#pragma GLOBAL_ASM("./asm/non_matchings/overlays/ovl_Obj_Warpstone_0x80B92B10/ObjWarpstone_Init.asm")
-
-#pragma GLOBAL_ASM("./asm/non_matchings/overlays/ovl_Obj_Warpstone_0x80B92B10/ObjWarpstone_Destroy.asm")
-
-#pragma GLOBAL_ASM("./asm/non_matchings/overlays/ovl_Obj_Warpstone_0x80B92B10/func_80B92C00.asm")
-
-#pragma GLOBAL_ASM("./asm/non_matchings/overlays/ovl_Obj_Warpstone_0x80B92B10/func_80B92C48.asm")
-
-#pragma GLOBAL_ASM("./asm/non_matchings/overlays/ovl_Obj_Warpstone_0x80B92B10/func_80B92CD0.asm")
-
-#pragma GLOBAL_ASM("./asm/non_matchings/overlays/ovl_Obj_Warpstone_0x80B92B10/func_80B92DC4.asm")
-
-#pragma GLOBAL_ASM("./asm/non_matchings/overlays/ovl_Obj_Warpstone_0x80B92B10/ObjWarpstone_Update.asm")
-
-#pragma GLOBAL_ASM("./asm/non_matchings/overlays/ovl_Obj_Warpstone_0x80B92B10/ObjWarpstone_Draw.asm")
+static Gfx* sOwlStatueDLs[] = { D_060001D0, D_06003770 };
+
+void ObjWarpstone_SetupAction(ObjWarpstone* this, ObjWarpstoneActionFunc actionFunc) {
+    this->actionFunc = actionFunc;
+}
+
+void ObjWarpstone_Init(Actor* thisx, GlobalContext* globalCtx) {
+    ObjWarpstone* this = THIS;
+
+    Actor_ProcessInitChain(&this->dyna.actor, sInitChain);
+    Collider_InitAndSetCylinder(globalCtx, &this->collider, &this->dyna.actor, &sCylinderInit);
+    Actor_SetHeight(&this->dyna.actor, 40.0f);
+
+    if (!IS_OWL_HIT(GET_OWL_ID(this))) {
+        ObjWarpstone_SetupAction(this, ObjWarpstone_ClosedIdle);
+    } else {
+        ObjWarpstone_SetupAction(this, ObjWarpstone_OpenedIdle);
+        this->modelIndex = SEK_MODEL_OPENED;
+    }
+}
+
+void ObjWarpstone_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    ObjWarpstone* this = THIS;
+
+    Collider_DestroyCylinder(globalCtx, &this->collider);
+}
+
+s32 ObjWarpstone_ClosedIdle(ObjWarpstone* this, GlobalContext* globalCtx) {
+    if (this->collider.base.acFlags & AC_HIT) {
+        ObjWarpstone_SetupAction(this, ObjWarpstone_BeginOpeningCutscene);
+        return true;
+    } else {
+        /*Ye who hold the sacred sword, leave proof of our encounter.*/
+        this->dyna.actor.textId = 0xC00;
+        return false;
+    }
+}
+
+s32 ObjWarpstone_BeginOpeningCutscene(ObjWarpstone* this, GlobalContext* globalCtx) {
+    if (this->dyna.actor.cutscene < 0 || ActorCutscene_GetCanPlayNext(this->dyna.actor.cutscene)) {
+        ActorCutscene_StartAndSetUnkLinkFields(this->dyna.actor.cutscene, &this->dyna.actor);
+        ObjWarpstone_SetupAction(this, ObjWarpstone_PlayOpeningCutscene);
+        Audio_PlayActorSound2(&this->dyna.actor, NA_SE_EV_OWL_WARP_SWITCH_ON);
+    } else {
+        ActorCutscene_SetIntentToPlay(this->dyna.actor.cutscene);
+    }
+    return true;
+}
+
+s32 ObjWarpstone_PlayOpeningCutscene(ObjWarpstone* this, GlobalContext* globalCtx) {
+    if (this->timer++ >= OBJ_WARPSTONE_TIMER_DONE) {
```

Comment:
If possible we tend to prefer keeping increments/decrements outside of conditionals assuming they still match

Reply:
People in my stream were telling me to put it inside the conditional so that's what I did, and it was the first thing I tried that matched. I can mess around and see if I can find a match outside the conditional.

Reply:
It should be as simple as moving the increment outside the conditional and just checking the timer. Not a huge deal if it doesn't work though. Having it outside is much more clear though

Reply:
Doesn't match (I also tried putting it inside the the if () {} block and that doesn't match either).

Reply:
Sorta makes sense if you think about it. If it was a prefix increment (++this->timer) then putting it before the conditional should work, but I had tried the prefix increment during my initial decomp and that didn't work either.

Reply:
I'm also not a huge fan of this timer macro since it's only a 1-time use macro, and if you're reading this code, then instead of the timer being right here where it's used, you have to go look it up

Reply:
It is only a one-time use enum, but it makes it much more obvious what the code is doing. Before I put it in, I had literally no idea the difference between the if, elseif, and else. Furthermore, I was told that one of the goals of decomp is to make things easy for modders in the future, and if a modder wanted to make owl statues take longer or shorter to open, the best way would be to just change the values in the enum in the header file. The code would still be just as legible, and they wouldn't even have to look at the .C file code at all for this particular task.

Reply:
Also getting rid of magic numbers in code is generally desirable afaik. Would rather use a constant defined elsewhere.

Reply:
Makes sense, the most important thing is to make it as clear as possible. There's a timer used in nearly every actor (in bigslime alone there's 26 timers used, one for basically each action it takes), so making timer `#defines` feels like it could open a giant rabbit hole of an endless amount of macros, but perhaps it's something we could consider. I think there would be two alternative approaches I would take instead of making a define for a macro timer limit, but these are entirely subjective and you're free to choose how you want to have it in the end, including just leaving it as is. I'd be fine with that, this is just friendly doc advise (which isn't even required atm).

1) In early round of docs, I would name the timer to something more descriptive. Something like `this->openingCSTimer` (I know I made the comments about abbreviations lol, but CS is short for cutscenes used in OoT).

2) In more thorough docs, you can add comments to if-else statements explaining what's going on and above the file, you have something like:
```
/**
 * Insert function description here
 * The if-else checks for this and that
 */
```
making things very clear what's going on in the function. But in the end, I'll leave it up to you

Reply:
renamed timer in latest push. marking as resolved.